File: leetcode/135-candy.cpp
/*
There are N children standing in a line. Each child is assigned a rating value.

You are giving candies to these children subjected to the following requirements:

Each child must have at least one candy.
Children with a higher rating get more candies than their neighbors.
What is the minimum candies you must give?
*/

class Solution {
public:
    int candy(vector<int>& ratings) {
        int n = ratings.size();
        vector<int> candies(n, 0);
        for (int i=0; i<n; ++i) {
            if ((i == 0 || ratings[i] <= ratings[i-1]) && (i == n-1 || ratings[i] <= ratings[i+1])) {
                candies[i] = 1;
                for (int j=i+1; j<n; ++j) {
                    if (ratings[j-1] < ratings[j]) {
                        candies[j] = max(candies[j], (ratings[j-1] == ratings[j] ? 0 : 1) + candies[j-1]);
                    } else {
                        break;
                    }
                }
                for (int j=i-1; j>=0; --j) {
                    if (ratings[j+1] < ratings[j]) {
                        candies[j] = max(candies[j], (ratings[j+1] == ratings[j] ? 0 : 1) + candies[j+1]);
                    } else {
                        break;
                    }
                }
            }
        }
        return accumulate(candies.begin(), candies.end(), 0);
    }
};

File: leetcode/135-candy.rb
# @param {Integer[]} ratings
# @return {Integer}
def candy(ratings)
  n = ratings.size
  candies = Array.new n, 1
  i = 0
  while i < n
    if (i == 0 and i < n-1 and ratings[i+1] >= ratings[i]) or (i == n-1 and i > 0 and ratings[i-1] >= ratings[i]) or (i > 0 and i < n-1 and ratings[i-1] >= ratings[i] and ratings[i+1] >= ratings[i]) or (i == 0 and i == n-1)
      j = 1
      while i - j >= 0 and ratings[i-j] > ratings[i-j+1]
        new_candies = candies[i-j+1] + 1
        candies[i-j] = [candies[i-j], new_candies].max
        j += 1
      end
      j = 1
      while i + j < n and ratings[i+j] > ratings[i+j-1]
        new_candies = candies[i+j-1] + 1
        candies[i+j] = [candies[i+j], new_candies].max
        j += 1
      end
      i += j
    else
      i += 1
    end
  end
  candies.inject(:+)
end
